airflow/timetables/events.py

```diff
@@ -0,0 +1,111 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import itertools
+from typing import Iterable, Optional
+
+import pendulum
+from pendulum import DateTime
+
+from airflow.timetables.base import DagRunInfo, DataInterval, TimeRestriction, Timetable
+
+
+class EventsTimetable(Timetable):
+    """
+    Timetable that schedules DAG runs at specific listed datetimes. Suitable for
+    predictable but truly irregular scheduling such as sporting events.
+
+    :param event_dates: List of datetimes for the DAG to run at. Duplicates will be ignored.
+    :param restrict_to_events: Whether manual runs should use the most recent event or
+        the current time
+    :param description: A name for the timetable to display in the UI. Default None will be shown as
+                        "X Events" where X is the len of event_dates
+    """
+
+    def __init__(
+        self,
+        event_dates: Iterable[DateTime],
+        restrict_to_events: bool = False,
+        presorted: bool = False,
```

Comment:
We probably need to do this another way (e.g. maybe mandate the list must be presorted?) since this argument would show in the documentation and confuse users.

Reply:
Now I think of it large event table would have pretty bad performance anyway because the `dropwhile` would get slower and slower over time. We’ll probably need some better bisect logic in here anyway, so let’s get this in first and improve things later. Let’s add documentation saying this argument is an internal thing that should not be passed directly by the user (for now; we can change that later if needed).